Move `border-bottom-color` transition into `:hover`

koddsson opened the pull request:
In certain loading situations, we can transition from the initial border color, [which is set to the `color` attribute of the element](https://drafts.csswg.org/css-backgrounds-3/#propdef-border-bottom-color), to `transparent`. This transition is a not so good looking flash of all underline navigation items when the page load initially (see https://github.com/github/github/issues/155029 & https://github.com/github/github/issues/166084).

This PR fixes the issue by not setting an initial `transition` rule for underline navigation items but only on hover. The drawbacks of this approach are that there is no transition when users stop hovering the item but this was accepted as a good tradeoff.

See https://github.com/github/github/pull/167395 for this same fix implemented as a "hack". I figured an upstream PR would be more approriate.

/cc @primer/ds-core

simurai commented on src/navigation/underline-nav.scss:
```diff
@@ -23,16 +23,14 @@
     outline-offset: -1px;
-    transition-timing-function: ease-out;
-    transition-duration: 0.12s;
+    transition: border-bottom-color 0.36s ease-in;
```
Maybe we could keep the shorter transition timing to still make it feel "snappy" on hover?

```suggestion
    transition: border-bottom-color 0.12s ease-out;
```

simurai reviewed (commented):
🙇  Thanks for the upstream.

~~I'll change the base branch to the one we use for the color modes so we don't have to wait for a new Primer CSS release and can move it to dotcom a lot earlier.~~

I reverted changing the base branch since there are merge conflicts that I try to first resolve.

koddsson commented on src/navigation/underline-nav.scss:
```diff
@@ -22,8 +22,7 @@
   // stylelint-disable-next-line primer/borders
-  border-bottom: 2px $border-style transparent;
-  transition: border-bottom-color 0.36s ease-in;
+  border-bottom: 2px $border-style rgba($border-gray-dark, 0);
```
Fix this after the merge

```suggestion
  border-bottom: 2px $border-style transparent;
```

koddsson commented:
> I reverted changing the base branch since there are merge conflicts that I try to first resolve.

I changed the base branch and resolved the conflicts! Here's how the navigation looks in this PR.

https://user-images.githubusercontent.com/318208/105017678-f5e61900-5a3b-11eb-9f85-ba334125f860.mov

simurai reviewed (approved):
> I changed the base branch and resolved the conflicts! Here's how the navigation looks in this PR.

👍 

Ok, thanks. 🙇  Let's merge this and I'll update `@primer/css-next` so that this change will be part of https://github.com/github/github/pull/165690 on dotcom.
